### PR TITLE
Readme and Examples for Tensor mode

### PR DESCRIPTION
**Pull request by YWolfeee**
Add a readme file for the tensor fitting, and add a dipole example and a polarizability example.

**Author** YWolfeee
**Date** May, 4, 2021

### CODE REVIEW

File: doc/Train a Deep Potential model to fit tensor like Dipole and Polarizability.md [hunk 1]
```diff
@@ -0,0 +1,109 @@
+# Train a Deep Potential model to fit `tensor` like `Dipole` and `Polarizability`
+
+Unlike `energy` which is a scalar, one may want to fit some high dimensional physical quantity, like `dipole` (vector) and `polarizability` (matrix, shorted as `polar`). Deep Potential has provided different API to allow this. In this example we will show you how to train a model to fit them for a water system. A complete training input script of the examples can be found in 
+
+```bash
+$deepmd_source_dir/examples/water_tensor/dipole/dipole_input.json
+$deepmd_source_dir/examples/water_tensor/polar/polar_input.json
+```
+
+Relevant data are also provided in our examples. But note that **our examples are of very limited amount, and should not be used to train a productive model.**
```
Comment (amcadmus) [inline doc/Train a Deep Potential model to fit tensor like Dipole and Polarizability.md:10]: "Relevant data are also provided" -> "The training and validation data are also provided..." 

"our examples are of very limited amount" -> "The data provided along with the examples are of limited amount"

File: doc/Train a Deep Potential model to fit tensor like Dipole and Polarizability.md [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+Relevant data are also provided in our examples. But note that **our examples are of very limited amount, and should not be used to train a productive model.**
+
+
+
+The directory of this examples:
+
+-   [The training input script](## the-training-input-script)
+-   [Training Data Preparation](## training-data-preparation)
+- 	[Train the Model](## train-the-model)
```
Comment (amcadmus) [inline doc/Train a Deep Potential model to fit tensor like Dipole and Polarizability.md:18]: The hyper links are broken, plz fix them.

File: doc/Train a Deep Potential model to fit tensor like Dipole and Polarizability.md [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+Similar to the `input.json` used in `ener` mode, training json is also divided into `model`, `learning_rate`, `loss` and `training`. Most `args` remains the same as `ener` mode, and their meaning can be found [here](https://github.com/deepmodeling/deepmd-kit/blob/devel/doc/train-se-e2-a.md). To fit a tensor, one need to modify `model.fitting_net` and `loss`.
+
+### Model
+
+The `fitting_net` section tells DP which fitting net to use, and should be provided like
+
+```json
+	"fitting_net" : {
+	    "type": "dipole" / "polar",
```
Comment (amcadmus) [inline doc/Train a Deep Potential model to fit tensor like Dipole and Polarizability.md:30]: This is not a valid json syntax. 
I would suggest to give two pieces of examples one for `dipole` and one for `polar`

File: doc/Train a Deep Potential model to fit tensor like Dipole and Polarizability.md [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+The directory of this examples:
+
+-   [The training input script](## the-training-input-script)
+-   [Training Data Preparation](## training-data-preparation)
+- 	[Train the Model](## train-the-model)
+
+## The training input script
+
+Similar to the `input.json` used in `ener` mode, training json is also divided into `model`, `learning_rate`, `loss` and `training`. Most `args` remains the same as `ener` mode, and their meaning can be found [here](https://github.com/deepmodeling/deepmd-kit/blob/devel/doc/train-se-e2-a.md). To fit a tensor, one need to modify `model.fitting_net` and `loss`.
```
Comment (amcadmus) [inline doc/Train a Deep Potential model to fit tensor like Dipole and Polarizability.md:22]: please don't use absolute path like `https://github.com/deepmodeling/deepmd-kit/blob/devel/doc/train-se-e2-a.md`. relative position `train-se-e2-a.md` should work

Comment (amcadmus) [inline doc/Train a Deep Potential model to fit tensor like Dipole and Polarizability.md:22]: " Most `args` remains" ->"Most keywords remain`

File: doc/Train a Deep Potential model to fit tensor like Dipole and Polarizability.md [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+	"fitting_net" : {
+	    "type": "dipole" / "polar",
+		"sel_type": [0],
+		"neuron": [100,100,100],
+		"resnet_dt": true,
+		"seed": 1,
+	},
+```
+
+-   `type` specifies which type of fitting net should be used. It should be either `dipole` or `polar`. Note that `global_polar` mode is already **deprecated** and is merged into `polar`. To specify whether a system is global or atomic, please see **here**.
```
Comment (amcadmus) [inline doc/Train a Deep Potential model to fit tensor like Dipole and Polarizability.md:38]: " Note that `global_polar` mode " ->" Note that `global_polar` mode in version 1.x"

File: doc/Train a Deep Potential model to fit tensor like Dipole and Polarizability.md [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+-   `sel_type` is a list specifying which type of atoms have the quantity you want to fit. For example, in water system, `sel_type` is `[0]` since `0` represents for atom `O`. If left unset, all type of atoms will be fitted.
+-   The rest `args` has the same meaning as they do in `ener` mode.
+
+### Loss
+
+DP supports a combinational training of global system (only a global `tensor` label, i.e. dipole or polar, is provided in a frame) and atomic system (labels for **each** atom included in `sel_type` are provided). The `loss` section tells DP the weight of this two kind of loss, i.e.
```
Comment (amcadmus) [inline doc/Train a Deep Potential model to fit tensor like Dipole and Polarizability.md:44]: We should explain what do we mean by "atomic tensor label" and by "global tensor label"

**Review by amcadmus — CHANGES_REQUESTED** [review]
` doc/Train a Deep Potential model to fit tensor like Dipole and Polarizability.md` is not filename that compatible to the naming convention of  doc files. Something like `doc/train-tensor.md` is preferred

File: doc/train-fitting-tensor.md [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+
+The training and validation data are also provided our examples. But note that **the data provided along with the examples are of limited amount, and should not be used to train a productive model.**
+
+
+
+The directory of this examples:
+
+-   [The training input script](# the-training-input-script)
+-   [Training Data Preparation](# training-data-preparation)
+- 	[Train the Model](# train-the-model)
```
Comment (amcadmus) [inline doc/train-fitting-tensor.md:18]: The hyperlink is not working
plz refer to https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#links for help

Please make sure the links are working before PR, thanks!

File: doc/train-fitting-tensor.md [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+		"seed": 1,
+	},
+```
+
+The json of `polar` type should be provided like
+
+```json
+"fitting_net" : {
+    "type": "polar",
+	"sel_type": [0],
```
Comment (amcadmus) [inline doc/train-fitting-tensor.md:45]: The indent should be kept consistent

File: doc/train-fitting-tensor.md [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+### Model
+
+The `fitting_net` section tells DP which fitting net to use.
+
+The json of `dipole` type should be provided like
+
+```json
+	"fitting_net" : {
+	    "type": "dipole",
+		"sel_type": [0],
```
Comment (amcadmus) [inline doc/train-fitting-tensor.md:33]: The indent should be kept consistent

File: doc/train-fitting-tensor.md [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+-   `sel_type` is a list specifying which type of atoms have the quantity you want to fit. For example, in water system, `sel_type` is `[0]` since `0` represents for atom `O`. If left unset, all type of atoms will be fitted.
+-   The rest `args` has the same meaning as they do in `ener` mode.
+
+### Loss
+
+DP supports a combinational training of global system (only a global `tensor` label, i.e. dipole or polar, is provided in a frame) and atomic system (labels for **each** atom included in `sel_type` are provided). In a global system, each frame has just **one** `tensor` label. For example, when fitting `polar`, each frame will just provide a `1 x 9` vector indicating the summation of polarizability. By contrast, in a atomic system, each atom in `sel_type` has a `tensor` label. For example, when fitting dipole, each frame will provide a `#sel_atom x 3` matrix, where `#sel_atom` is the number of atoms whose type are in `sel_type`.
```
Comment (amcadmus) [inline doc/train-fitting-tensor.md:58]: "... provide a `1 x 9` vector indicating the summation of polarizability" -> 
"... provide a `1 x 9` vector which gives the elements of the polarizability tensor of that frame in order XX, XY, XZ, YX, YY, YZ, XZ, ZY, ZZ."

File: doc/train-fitting-tensor.md [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+  1400     3.13e-02   5.76e-02   3.13e-02   5.76e-02    0.00e+00    0.00e+00   6.3e-07
+  1500     3.34e-02   1.11e-02   2.09e-02   0.00e+00    1.57e-03    1.39e-03   3.2e-07
+  1600     3.11e-02   5.64e-02   3.11e-02   5.64e-02    0.00e+00    0.00e+00   1.6e-07
+  1700     2.97e-02   5.05e-02   2.97e-02   5.05e-02    0.00e+00    0.00e+00   7.9e-08
+  1800     2.64e-02   7.70e-02   1.09e-02   0.00e+00    1.94e-03    9.62e-03   4.0e-08
+  1900     3.28e-02   2.56e-02   3.28e-02   0.00e+00    0.00e+00    3.20e-03   2.0e-08
+  2000     2.59e-02   5.71e-02   1.03e-02   5.71e-02    1.94e-03    0.00e+00   1.0e-08
+```
+
+One may notice that in each step, some of local loss and global loss will be `0.0`, which is normal since we just select some systems to train in each time step, and it's possible that all data we select are from one type of system. Nevertheless, `rmse_val` and `rmse_trn` is still meaningful.
```
Comment (amcadmus) [inline doc/train-fitting-tensor.md:124]: This explanation is not very clear. We may give an example of the `training_data` and `validation_data` section, and tell the users some of the systems only have global labels while some of them only have atomic labels. During training, at each step when the lcurve.out is printed, the system used for evaluating the training (validation) error may be either with only global or only atomic labels, thus the corresponding atomic or global errors are missing and are printed as zeros.

**Review by amcadmus — CHANGES_REQUESTED** [review]
